### PR TITLE
[web] Don't reset history on hot restart

### PR DESCRIPTION
**Pull request by mdebbar**
When hot restart happens, instead of resetting browser history, we just remember which history mode was being used. Once the app starts again after the hot restart, we re-create the same history mode. Then:

1. If the newly started app uses the same history mode: nothing will happen. The app will just start from where it left off.
2. If the newly started app uses a different history mode: the previous history mode will be asked to tear down, then the new mode will setup itself.

Fixes https://github.com/flutter/flutter/issues/79241

### CODE REVIEW

File: lib/web_ui/lib/src/engine/window.dart [hunk 1]
```diff
@@ -49,11 +77,15 @@ class EngineFlutterWindow extends ui.SingletonFlutterWindow {
     engineDispatcher.windows[_windowId] = this;
     engineDispatcher.windowConfigurations[_windowId] = const ui.ViewConfiguration();
     if (_isUrlStrategySet) {
-      _browserHistory =
-          MultiEntriesBrowserHistory(urlStrategy: _customUrlStrategy);
+      _browserHistory = _createDefaultBrowserHistory(_customUrlStrategy);
     }
     registerHotRestartListener(() {
```
Comment (chunhtai) [inline lib/web_ui/lib/src/engine/window.dart:82]: Does the hot restart teardown the history entries? If it does, we will lose all the url and state even if we memorize the mode. If it does not, I think we can just read the current state to determine whether that is a multi entry or single entry

**Review by chunhtai — COMMENTED** [review]
We need a test, the change lgtm

**Comment by mdebbar** [comment]
I'm relying on existing tests to catch any regression caused by this PR.

I also filed an issue to add an integration test. I need to figure out how to test hot restart on web.

**Comment by chunhtai** [comment]
sounds good, I think we can write a simple test for createHistoryForExistingState before we have integration test coverage

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**